service/storage_service.cc

```diff
@@ -2773,22 +2774,245 @@ void storage_service::add_expire_time_if_found(inet_address endpoint, int64_t ex
+                failed = true;
+                eptr = std::current_exception();
+                slogger.warn("load_and_stream: ops_uuid={}, ks={}, table={}, del_sst_phase, err={}",
+                        ops_uuid, ks_name, cf_name, eptr);
+            }
+        }
+        auto duration = std::chrono::duration_cast<std::chrono::duration<float>>(std::chrono::steady_clock::now() - start_time).count();
+        for (auto& [node, meta] : metas) {
+            slogger.info("load_and_stream: ops_uuid={}, ks={}, table={}, target_node={}, num_partitions_sent={}, num_bytes_sent={}",
+                    ops_uuid, ks_name, cf_name, node, meta.num_partitions_sent(), meta.num_bytes_sent());
+        }
+        auto partition_rate = std::fabs(duration) > FLT_EPSILON ? num_partitions_processed / duration : 0;
+        auto bytes_rate = std::fabs(duration) > FLT_EPSILON ? num_bytes_read / duration / 1024 / 1024 : 0;
+        auto status = failed ? "failed" : "succeeded";
+        slogger.info("load_and_stream: finished ops_uuid={}, ks={}, table={}, partitions_processed={} partitions, bytes_processed={} bytes, partitions_per_second={} partitions/s, bytes_per_second={} MiB/s, duration={} s, status={}",
+                ops_uuid, ks_name, cf_name, num_partitions_processed, num_bytes_read, partition_rate, bytes_rate, duration, status);
+        if (failed) {
+            std::rethrow_exception(eptr);
+        }
+    }
+    co_return;
+}
+
 // For more details, see the commends on column_family::load_new_sstables
 // All the global operations are going to happen here, and just the reloading happens
 // in there.
-future<> storage_service::load_new_sstables(sstring ks_name, sstring cf_name) {
+future<> storage_service::load_new_sstables(sstring ks_name, sstring cf_name,
+    bool load_and_stream, bool primary_replica_only) {
     if (_loading_new_sstables) {
         throw std::runtime_error("Already loading SSTables. Try again later");
     } else {
         _loading_new_sstables = true;
     }
-
-    slogger.info("Loading new SSTables for {}.{}...", ks_name, cf_name);
-
-    return distributed_loader::process_upload_dir(_db, _sys_dist_ks, _view_update_generator, ks_name, cf_name).finally([this, ks_name, cf_name] {
-        slogger.info("Done loading new SSTables for {}.{}", ks_name, cf_name);
+    slogger.info("Loading new SSTables for keyspace={}, table={}, load_and_stream={}, primary_replica_only={}",
+            ks_name, cf_name, load_and_stream, primary_replica_only);
+    try {
+        if (load_and_stream) {
+            utils::UUID table_id;
+            std::vector<std::vector<sstables::shared_sstable>> sstables_on_shards;
+            std::tie(table_id, sstables_on_shards) = co_await distributed_loader::get_sstables_from_upload_dir(_db, ks_name, cf_name);
+            co_await service::get_storage_service().invoke_on_all([&sstables_on_shards, ks_name, cf_name, table_id, primary_replica_only] (storage_service& ss) mutable -> future<> {
```

Comment:
storage_service is a `peering_sharded_service`, so you can use container() instead of service::get_storage_service().

Reply:
OK

Reply:
Done. Fixed in v4.